src/Backend/Modules/Tags/Engine/Model.php

```diff
@@ -289,79 +320,100 @@ public static function saveTags($otherId, $tags, $module, $language = null)
-                // get tagId
-                $tagId = (int) $tagsAndIds[$tag];
-
-                // not linked before so increment the counter
-                if (!isset($currentTags[$tag])) {
-                    $db->execute(
-                        'UPDATE tags SET number = number + 1 WHERE id = ?',
-                        $tagId
-                    );
-                }
+                // tag doesn't exist yet
+                if (!isset($existingTagIds[$tag])) {
+                    // build new tag
+                    $item = new Tag();
+                    $item
+                        ->setTag($tag)
+                        ->setNumber(0)
+                        ->setUrl(self::getURL($tag))
+                        ->setLanguage($language)
+                    ;
 
-                // add to insert array
-                $rowsToInsert[] = array('module' => $module, 'tag_id' => $tagId, 'other_id' => $otherId);
+                    // insert tag
+                    self::insert($item);
+
+                    // add to tags to connect
+                    $tagsToConnect[] = $item;
+                }
             }
 
-            // insert the rows at once if there are items to insert
-            if (!empty($rowsToInsert)) {
-                $db->insert('modules_tags', $rowsToInsert);
+            // loop all tags to connect
+            foreach ($tagsToConnect as $tag) {
+                // build new tag connection
+                $item = new TagConnection();
+                $item
+                    ->setTag($tag)
+                    ->setModule($module)
+                    ->setOtherId($otherId)
+                ;
+
+                // add connection
+                $tag->addConnection($item);
+
+                // bump number
+                $tag->setNumber($tag->getNumber() + 1);
+
+                // update
+                $em->persist($item);
+                $em->persist($tag);
             }
+            $em->flush();
         }
 
         // add to search index
-        BackendSearchModel::saveIndex($module, $otherId, array('tags' => implode(' ', (array) $tags)), $language);
-
-        // decrement number
-        foreach ($currentTags as $tag => $tagId) {
-            // if the tag can't be found in the new tags we lower the number of tags by one
-            if (array_search($tag, $tags) === false) {
-                $db->execute(
-                    'UPDATE tags SET number = number - 1 WHERE id = ?',
-                    $tagId
-                );
-            }
-        }
+        BackendSearchModel::saveIndex(
+            $module,
+            $otherId,
+            array(
+                'tags' => implode(' ', (array) $tags)
+            ),
+            $language
+        );
 
         // remove all tags that don't have anything linked
-        $db->delete('tags', 'number = ?', 0);
+        $em
+            ->createQuery('delete from ' . self::ENTITY_CLASS . ' i where i.number = 0')
```

Comment:
we write SQL keywords uppercased. So `DELETE FROM' . self::ENTITY_CLASS . ' i WHERE i.number = 0`